Add system catalog dumper for self-managed debug tool

SangJunBak opened the pull request:
- See commit messages for details

### Motivation

  * This PR adds a known-desirable feature.
https://github.com/MaterializeInc/database-issues/issues/8908

<!--
Which of the following best describes the motivation behind this PR?

  * This PR fixes a recognized bug.

    [Ensure issue is linked somewhere.]


    [Ensure issue is linked somewhere.]

  * This PR fixes a previously unreported bug.

    [Describe the bug in detail, as if you were filing a bug report.]

  * This PR adds a feature that has not yet been specified.

    [Write a brief specification for the feature, including justification
     for its inclusion in Materialize, as if you were writing the original
     feature specification.]

   * This PR refactors existing code.

    [Describe what was wrong with the existing code, if it is not obvious.]
-->

### Tips for reviewer

<!--
Leave some tips for your reviewer, like:

    * The diff is much smaller if viewed with whitespace hidden.
    * [Some function/module/file] deserves extra attention.
    * [Some function/module/file] is pure code movement and only needs a skim.

Delete this section if no tips.
-->

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

SangJunBak commented:
I just realized it's probably a bad idea to expose tracing as stdout/stderr... 

@ParkMyCar Should we revert back to `println/eprintln`? At the end, I also want to take everything we printed to stdout, stderr, and maybe some additional debug info, then include that as a file in the debug zip.

ParkMyCar commented on src/self-managed-debug/src/system_catalog_dumper.rs:
```diff
@@ -206,3 +217,577 @@ pub fn spawn_sql_port_forwarding_process(
+
+#[derive(Debug, Clone)]
+pub struct ClusterReplica {
+    // TODO (debug_tool1): Use this when scraping introspection relations
```
love the use of these TODOs!

ParkMyCar commented on src/self-managed-debug/src/system_catalog_dumper.rs:
```diff
@@ -206,3 +217,577 @@ pub fn spawn_sql_port_forwarding_process(
+
+impl<'n> SystemCatalogDumper<'n> {
+    pub async fn new(context: &'n Context, connection_string: &str) -> Result<Self, anyhow::Error> {
+        let (pg_client, pg_conn) = retry::Retry::default()
```
FWIW we have our own `PostgresClient` [here](https://github.com/MaterializeInc/materialize/blob/main/src/postgres-client/src/lib.rs#L94-L96) that handles a lot of nice things like connection pooling.

Not necessary to use now but might be worth taking a look!

ParkMyCar commented on src/self-managed-debug/src/system_catalog_dumper.rs:
```diff
@@ -206,3 +217,577 @@ pub fn spawn_sql_port_forwarding_process(
+                "SET search_path = mz_internal, mz_catalog, mz_introspection",
+                &[],
+            )
+            .await?;
```
would be nice to add some `.context("set search path")` here, and to the other DB calls! Otherwise when we get an error back we don't know which query caused the error to fail

ParkMyCar commented on src/self-managed-debug/src/system_catalog_dumper.rs:
```diff
@@ -206,3 +217,577 @@ pub fn spawn_sql_port_forwarding_process(
+        // Set search path to system catalog tables
+        pg_client
+            .execute(
+                "SET search_path = mz_internal, mz_catalog, mz_introspection",
```
Slight preference to moving these queries to a static string, e.g.
```
static SET_SEARCH_PATH_QUERY: &str = "SET search_path = mz_internal, mz_catalog, mz_introspection";
```
but up to you!

ParkMyCar commented on src/self-managed-debug/src/system_catalog_dumper.rs:
```diff
@@ -206,3 +217,577 @@ pub fn spawn_sql_port_forwarding_process(
+            {
+                Ok(rows) => rows.into_iter()
+                            .map(|row| ClusterReplica {
+                                _cluster_name: row.get::<_, String>("cluster_name"),
+                                _replica_name: row.get::<_, String>("replica_name"),
```
`get(...)` panics if the column doesn't exist, probably better to use `try_get(...)` here! e.g. something like this should get you close:
```
rows.into_iter()
    .map(|row| Ok(ClusterReplica {
        _cluster_name: row.try_get::<_, String>("cluster_name")?,
        _replica_name: row.try_get::<_, String>("replica_name")?,
    }))
    .collect::<Result<Vec<_>, _>>(),
```

ParkMyCar commented on src/self-managed-debug/src/system_catalog_dumper.rs:
```diff
@@ -206,3 +217,577 @@ pub fn spawn_sql_port_forwarding_process(
+                            }
+                            _ => {}
+                        }
+                        // TODO: Change None based on relation category
```
nit, use `TODO (debug_tool)` here

ParkMyCar commented on src/self-managed-debug/src/system_catalog_dumper.rs:
```diff
@@ -206,3 +217,577 @@ pub fn spawn_sql_port_forwarding_process(
+}
+
+const RELATIONS: &[Relation] = &[
+    // Basic object information
```
nit: can you move this to a doc comment on the `const RELATIONS`?

ParkMyCar commented on src/self-managed-debug/src/system_catalog_dumper.rs:
```diff
@@ -206,3 +217,577 @@ pub fn spawn_sql_port_forwarding_process(
+    pub category: RelationCategory,
+}
+
+const RELATIONS: &[Relation] = &[
```
`static` probably works better here instead of `const`.

When something is `const` the compiler will inline the entire list wherever it's used, so this list will get copied around a bunch. When it's `static` we'll have a single copy in static memory that we pass around references to.

ParkMyCar commented on src/self-managed-debug/src/system_catalog_dumper.rs:
```diff
@@ -206,3 +217,577 @@ pub fn spawn_sql_port_forwarding_process(
+        name: "mz_object_transitive_dependencies",
+        category: RelationCategory::Basic,
+    },
+    // Compute
```
ohhhh I see now, these comments categorize the types of relations we're querying. Mind calling _that_ out in a doc comment so when folks add new relations they know they're sorted?

ParkMyCar commented on src/self-managed-debug/src/system_catalog_dumper.rs:
```diff
@@ -206,3 +217,577 @@ pub fn spawn_sql_port_forwarding_process(
+    }
+}
+
+pub fn create_postgres_connection_string(
```
Can you document the defaults that get used if params are not provided?

ParkMyCar commented on src/self-managed-debug/src/utils.rs:
```diff
@@ -0,0 +1,24 @@
+
+use chrono::{DateTime, Utc};
+
+pub fn format_base_path(date_time: DateTime<Utc>) -> PathBuf {
```
What does base_path mean in this context? I'm guessing it's where we write the data dump to? Would be great to document

ParkMyCar commented on src/self-managed-debug/src/k8s_resource_dumper.rs:
```diff
@@ -366,13 +367,8 @@ fn format_resource_path(
-        MAIN_SEPARATOR,
-        resource_type,
-    ));
+    let mut path = format_base_path(date_time).join(resource_type);
```
Not necessarily in this PR, but just realizing now, what happens if this path already exists?

SangJunBak replied:
Ah so at the very end, once the full path is composed. i.e. in `system_catalog_dumper` lines 766 to 768:
```
            let file_path = format_file_path(start_time, None);
            let file_path_name = file_path.join(relation_name).with_extension("csv");
            tokio::fs::create_dir_all(&file_path).await?;
```

We do a `create_dir_all` which recursively creates the `dir` if missing:

<img width="687" alt="Screenshot 2025-03-20 at 12 41 04 AM" src="https://github.com/user-attachments/assets/be3a9bb9-024e-4018-93c7-8b57f383c73c" />

SangJunBak commented:
@ParkMyCar Thanks for the review! Addressed all your feedback. but in https://github.com/MaterializeInc/materialize/pull/31951 from commit  [Annotate postgresql query errors](https://github.com/MaterializeInc/materialize/pull/31951/commits/317bcdac40b5d7bd66318e139a3b9da0ee50246c) and onwards if that's cool! For the tracing feedback you gave, I actually wanna print out to the user what I'm currently tracing! But it's good to know that I can just change the tracing subscriber logic. I'll most likely clean it up when I'm implementing the zipping at the end!

At the end, I'll merge both PRs into a single PR then squash & commit!
